File: angel-ps/psf/src/main/java/com/tencent/angel/ml/matrix/psf/update/enhance/FullUpdateFunc.java
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tencent.angel.ml.matrix.psf.update.enhance;

import com.tencent.angel.PartitionKey;
import com.tencent.angel.ps.impl.PSContext;
import com.tencent.angel.ps.impl.matrix.ServerDenseDoubleRow;
import com.tencent.angel.ps.impl.matrix.ServerPartition;

/**
 * `FullUpdateFunc` is a PSF Update for the whole matrix in matrix with a user-defined function.
 */

public abstract class FullUpdateFunc extends UpdateFunc {
  public FullUpdateFunc(int matrixId, double[] values) {
    super(new FullUpdateParam(matrixId, values));
  }

  public FullUpdateFunc() {
    super(null);
  }

  @Override
  public void partitionUpdate(PartitionUpdateParam partParam) {
    ServerPartition part = psContext
        .getMatrixStorageManager()
        .getPart(partParam.getMatrixId(), partParam.getPartKey().getPartitionId());

    if (part != null) {
      FullUpdateParam.FullPartitionUpdateParam ff =
          (FullUpdateParam.FullPartitionUpdateParam) partParam;

      update(part, partParam.getPartKey(), ff.getValues());
    }
  }

  private void update(ServerPartition part, PartitionKey key, double[] values) {
    switch (part.getRowType()) {
      case T_DOUBLE_DENSE:
        int startRow = key.getStartRow();
        int endRow = key.getEndRow();
        ServerDenseDoubleRow[] denseRows = new ServerDenseDoubleRow[endRow - startRow];
        for (int i = startRow; i < endRow; i++) {
          denseRows[i - startRow] = (ServerDenseDoubleRow) part.getRow(i);
        }
        doUpdate(denseRows, values);
        return;
      default:
        throw new RuntimeException("currently only supports DoubleDenseRow");
    }
  }

  protected abstract void doUpdate(ServerDenseDoubleRow[] rows, double[] values);

}

File: angel-ps/psf/src/main/java/com/tencent/angel/ml/matrix/psf/update/enhance/FullUpdateFunc.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in 
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */


package com.tencent.angel.ml.matrix.psf.update.enhance

import com.tencent.angel.PartitionKey
import com.tencent.angel.exception.AngelException
import com.tencent.angel.ml.matrix.RowType
import com.tencent.angel.ml.matrix.psf.update.base.{PartitionUpdateParam, UpdateFunc}
import com.tencent.angel.ps.storage.matrix.ServerPartition
import com.tencent.angel.ps.storage.vector._


/**
  * `FullUpdateFunc` is a PSF Update for the whole matrix in matrix with a user-defined function.
  */

abstract class FullUpdateFunc(param: FullUpdateParam) extends UpdateFunc(param) {
  def this(matrixId: Int, values: Array[Double]) = this(new FullUpdateParam(matrixId, values))

  def this() = this(null)

  override def partitionUpdate(partParam: PartitionUpdateParam) {
    val part = psContext.getMatrixStorageManager
      .getPart(partParam.getMatrixId, partParam.getPartKey.getPartitionId)
    if (part != null) {
      val ff = partParam.asInstanceOf[FullUpdateParam.FullPartitionUpdateParam]
      update(part, partParam.getPartKey, ff.getValues);
    }
  }

  private def update(part: ServerPartition, key: PartitionKey, values: Array[Double]) {
    val startRow = key.getStartRow
    val endRow = key.getEndRow
    part.getRowType match {
      case RowType.T_DOUBLE_DENSE =>
        doUpdate(Array.range(startRow, endRow).map(part.getRow(_).asInstanceOf[ServerIntDoubleRow]), values)
      case RowType.T_FLOAT_DENSE =>
        doUpdate(Array.range(startRow, endRow).map(part.getRow(_).asInstanceOf[ServerIntFloatRow]), values.map(_.toFloat))
      case RowType.T_LONG_DENSE =>
        doUpdate(Array.range(startRow, endRow).map(part.getRow(_).asInstanceOf[ServerIntLongRow]), values.map(_.toLong))
      case RowType.T_INT_DENSE =>
        doUpdate(Array.range(startRow, endRow).map(part.getRow(_).asInstanceOf[ServerIntIntRow]), values.map(_.toInt))
      case other => throw new AngelException(s"not supported operation for rowType = $other")
    }
  }

  def doUpdate(rows: Array[ServerIntDoubleRow], values: Array[Double])

  def doUpdate(rows: Array[ServerIntFloatRow], values: Array[Float])

  def doUpdate(rows: Array[ServerIntLongRow], values: Array[Long])

  def doUpdate(rows: Array[ServerIntIntRow], values: Array[Int])
}
